### PR TITLE
feat: default handling of decimal and Float64

### PR DESCRIPTION
**Pull request by Rishabh-V**
Default handling of decimal and Float64 with unit tests.
 
The implementation uses string splitting and parsing and makes use of dictionary to keep strongly typed mappings. 

Commit 1: Decimal and Float64 default handling.
Commit 2: Builds on top of Commit 1 and adds support for default DateTime and Date handling.

Will raise a new PR for integration tests once this PR is good.

### CODE REVIEW

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 1]
```diff
@@ -989,7 +989,128 @@ public async Task CanExecuteReadPartitionedReadCommand()
...
+            {
+                builder.SpannerToClrTypeDefaultMappings = spannerToClrTypeMapping;
+            }
+            var options = SpannerConversionOptions.ForConnectionStringBuilder(builder);
+            Value protobufValue = RunReadRequest(dbType, dbValue).Values[0];
+            // This is a hack, but simpler to do what DataReader does behind the scenes.
+            var clrValue = dbType.ConvertToClrType<object>(protobufValue, options);
+            // Check that CLR type of value is as expected.
+            Assert.Equal(clrType, clrValue.GetType());
+            // Check the values.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:1059]: This only checks the value for double. To be honest, I think it's probably simpler to just remove this bit and assume that if we've got the right type, the right conversion has taken place - that's tested elsewhere already.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 2]
```diff
@@ -989,7 +989,128 @@ public async Task CanExecuteReadPartitionedReadCommand()
...
+
+            var command = connection.CreateReadCommand(
+                 "Foo", ReadOptions.FromColumns(new List<string> { "Col1" }),
+                 KeySet.FromParameters(new SpannerParameterCollection
+                     {
+                        { "p", dbType, value } 
+                     }));
+            using var reader = command.ExecuteReader();
+            Assert.True(reader.HasRows);
+            // Handle the case of Float64 for now.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: I think it would be simpler to just return the request instead of doing any validation here. Let the caller do the validation.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionStringBuilderTests.cs [hunk 3]
```diff
@@ -215,5 +215,82 @@ public void WithDatabase()
...
+            builder.SpannerToClrTypeDefaultMappings = "Float64ToDecimal";
+            // DbConnectionStringBuilder lower-cases keywords, annoyingly.
+            Assert.Equal("spannertoclrtypedefaultmappings=Float64ToDecimal", builder.ToString());
+            // Check that internal mapping dictionary is populated with user provided configuration.
+            Assert.True(builder.SpannerToClrTypeMappings.Count == 1);
+            Assert.Equal(typeof(decimal), builder.SpannerToClrTypeMappings[SpannerDbType.Float64]);
+        }
+
+        [Theory]        
+        [InlineData("DecimalToFloat64,DecimalToNumeric;")] // Multiple mappings for a type.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerConnectionStringBuilderTests.cs]: It's not clear to me from this test whether *just* additional whitespace is a problem, e.g. "DecimalToFloat64    ". Every test that includes whitespace is *also* broken in some other way.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 4]
```diff
@@ -140,6 +181,160 @@ private string ValidatedDataSource(string dataSource)
...
+            }
+
+            // Check multiples for each type.
+            // Currently, we have Decimal to check.
+            if (mappings.Count(j => j.StartsWith("Decimal", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Decimal to SpannerDbType are provided for ${nameof(ClrToSpannerTypeDefaultMappings)}");
+            }
+
+            // If we reach here all is well. Return the input as validation passed.
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:271]: You could argue that that's putting "the connection string representation of the options" into SpannerConversionOptions, but I think that's *probably* okay.

**Review by jskeet — COMMENTED** [review]
I've only reviewed the first commit, because I'm suggesting a significant change - basically moving quite a lot of the code out of SpannerConnectionStringBuilder to SpannerConverisonOptions. (I've also suggested a change in option name, which I should have caught earlier - sorry about that.) I haven't actually reviewed the changes later than SpannerConnectionStringBuilder as they'd probably change if we moved things anyway.

I'm not suggesting a change to the public API, so we *could* potentially merge this (after the other comments are addressed) and then refactor. See what you think.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:271]: I was thinking, if I can do the same for "UseClrDefaultForNull" as that is the only other property in SpannerConnectionStringBuilder that impacts the SpannerConversionOptions. And post that I could just return the builder.ConversionOptions in the factory methods ForConnection and ForConnectionStringBuilder.  Please let me know if this sounds fine?

[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:271]: Yes, that sounds fine to me - I do like the idea that everything is in one place and that obtaining the conversion options is trivial.

**Comment by Rishabh-V** [comment]
> I've only reviewed the first commit, because I'm suggesting a significant change - basically moving quite a lot of the code out of SpannerConnectionStringBuilder to SpannerConverisonOptions. (I've also suggested a change in option name, which I should have caught earlier - sorry about that.) I haven't actually reviewed the changes later than SpannerConnectionStringBuilder as they'd probably change if we moved things anyway.
> 
> I'm not suggesting a change to the public API, so we _could_ potentially merge this (after the other comments are addressed) and then refactor. See what you think.

Hi Jon. I have tried to implement the suggested changes. Currently, it has both decimal and DateTime changes (unfortunately :( as I was working on the same branch).  Just so that the core implementation gets reviewed and I get the feedback, I have raised the PR containing both. 

Once the implementation gets ok, I will raise 2 different PRs for merging, one for decimal and Float64 and one for DateTime and Date.  Hope it is fine.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 5]
```diff
@@ -989,7 +989,128 @@ public async Task CanExecuteReadPartitionedReadCommand()
...
+            command.Parameters.Add(parameter);
+            using var reader = command.ExecuteReader();
+            Assert.True(reader.HasRows);
+            spannerClientMock.Verify(client => client.ExecuteStreamingSql(
+                It.IsAny<ExecuteSqlRequest>(),
+                It.IsAny<CallSettings>()), Times.Once());
+            return request.Params;
+        }
+
+        private ListValue RunReadRequest(SpannerDbType dbType, object value)
```
[the remark this answers is not in the document]
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:1083]: Right Jon. I have used it in the caller as reader will use it. Please let me know if this needs to be changed.

[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:1083]: Ah, so this isn't for testing the format of the parameter being sent, but the value being received? That's a little confusing - I'd expected this to be testing "is the request created for a given parameter correct".

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 6]
```diff
@@ -455,6 +568,21 @@ private string GetValueOrDefault(string key, string defaultValue = "")
                 {
                     value = ValidatedDataSource((string)value);
                 }
+                if (string.Equals(keyword, ClrToSpannerTypeDefaultMappingsKeyword, StringComparison.OrdinalIgnoreCase))
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:571]: All of these can be `else if`, I think. We could also probably change it just to set the releated property to `(string) value` (and then return), as a way of avoiding duplication.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 7]
```diff
@@ -46,6 +108,196 @@ private SpannerConversionOptions(bool useDBNull)
...
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Float64 to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            if (mappings.Count(j => j.StartsWith("Date", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Date to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // We don't want to modify the current instance, so create a new dictionary.
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:194]: I suspect we can do this without a dictionary to be honest... but now the logic is in the "right" place, I'm happier to refactor later.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs [hunk 8]
```diff
@@ -603,23 +604,30 @@ private object ConvertToClrTypeImpl(Value wireValue, System.Type targetClrType,
...
                             $"Invalid Type conversion from {wireValue.KindCase} to {targetClrType.FullName}");
                 }
             }
             if (targetClrType == typeof(PgNumeric))
             {
-                if (TypeCode != TypeCode.Numeric || TypeAnnotationCode != TypeAnnotationCode.PgNumeric)
+                bool isPgNumeric = TypeAnnotationCode == TypeAnnotationCode.PgNumeric;
+                bool isFloat64 = TypeCode == TypeCode.Float64;
+
+                if (!isPgNumeric && !isFloat64)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:619]: Should this be `||` instead of `&&`? (At the moment, we wouldn't fail if it's a string with a PgNumeric type annotation code for example. That might never be feasible, but it feels like it's logically `||`.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs [hunk 9]
```diff
@@ -603,23 +604,30 @@ private object ConvertToClrTypeImpl(Value wireValue, System.Type targetClrType,
...
+                    throw new ArgumentException($"{targetClrType.FullName} can only be used for numeric results and only when working with a PostgreSQL dialect.");
                 }
                 switch (wireValue.KindCase)
                 {
                     case Value.KindOneofCase.NullValue:
                         return null;
                     case Value.KindOneofCase.StringValue:
                         return V1.PgNumeric.Parse(wireValue.StringValue);
+                    case Value.KindOneofCase.NumberValue:
+                        return V1.PgNumeric.FromDecimal(Convert.ToDecimal(wireValue.NumberValue));
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:630]: Ick - I can see why we've got this, but it's unfortunate. Maybe we need to add a `FromDouble` as well at some point, if only internally.

**Review by jskeet — COMMENTED** [review]
I don't have time for a full review before Monday, but the separation of concerns definitely looks a lot cleaner now, thanks! Will review properly on Monday.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs (hunk 8 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:619]: We need `&&` here as PgNumeric value by default comes as string and while reading a Float64 value as PgNumeric, we need to read the NumberValue :(

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 10]
```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
...
+            // Empty string or null is valid to unset the values.
+            if (string.IsNullOrEmpty(input))
+            {
+                // Set defaults.
+                options.SetSpannerToClrTypeDefaults();
+                return;
+            }
+
+            // Get all type mappings.
+            var mappings = input.Split(new char[] { ',', ';' }, StringSplitOptions.RemoveEmptyEntries);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:161]: I suspect we don't need `;` in the list here. (I don't *think* it should come through from the connection string builder.)
That would allow us to use `input.Split(',',StringSplitOptions.RemoveEmptyEntries)'`, avoiding allocating the extra array.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 11]
```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
...
+
+            // Validate that mappings are valid.
+            if (mappings == null || !mappings.Any())
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:164]: `mappings` will never be null here.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 12]
```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
...
+            // Validate that mapping matches one of the pre-defined values.
+            var invalidMappings = mappings.Except(AllowedSpannerToClrTypeMappings, StringComparer.OrdinalIgnoreCase);
+            if (invalidMappings.Any())
+            {
+                throw new ArgumentException($"The following mappings are invalid:'{string.Join(",", invalidMappings)}' for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // Check multiples for each type.
+            // Currently, we have Float64 and Date to check.
+            if (mappings.Count(j => j.StartsWith("Float64", StringComparison.OrdinalIgnoreCase)) > 1)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:178]: I think we might want to have a more "data-oriented" approach to this, grouping options by affected type - but we don't need to do that immediately. (I'm happy to send a refactoring PR.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 13]
```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
...
+            if (mappings.Count(j => j.StartsWith("Date", StringComparison.OrdinalIgnoreCase)) > 1)
+            {
+                throw new ArgumentException($"'{input}' is not a valid value as multiple mappings from Date to CLR type are provided for ${nameof(SpannerConnectionStringBuilder.SpannerToClrTypeDefaultMappings)}");
+            }
+
+            // If we reach here, all is well.
+            foreach (var mapping in mappings)
+            {
+                // Discard pattern, primarily to simplify code using switch expressions.
+                _ = mapping switch
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:192]: This will be case-sensitive, having done a case-insensitive comparison earlier.
We could either make the comparisons case-sensitive, or lower-case the input and expected values, or use if statements.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs [hunk 14]
```diff
@@ -46,6 +116,159 @@ private SpannerConversionOptions(bool useDBNull)
...
+            DateTimeToConfiguredSpannerType = SpannerDbType.Timestamp;
+        }
+
+        internal void SetSpannerToClrTypeDefaults()
+        {
+            DateToConfiguredClrType = typeof(DateTime);
+            Float64ToConfiguredClrType = typeof(double);
+        }
+
+        private void ValidateAndParseClrToSpannerTypeMappings(string input, SpannerConversionOptions options)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:219]: Ideally, I'd like to reduce the duplication between this and the other method - although it's entirely possible that abstracting the infrastructure will be larger than the current code.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs (hunk 8 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:619]: I'm afraid I don't understand - and in particular, this *is* a change to behavior.
Was there a bug before?

As a concrete example: we *won't* throw an exception here if the user has requested a `PgNumeric` conversion for a `TypeCode.Float64` value in a regular Spanner database... that seems wrong, given the error message. Surely we want to throw an exception if *either* we're not in Postgres *or* we don't have a numeric value.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs [hunk 15]
```diff
@@ -316,6 +317,10 @@ public static SpannerDbType FromClrType(System.Type type)
             {
                 return Bytes;
             }
+            if (type == typeof(SpannerDate))
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:320]: Was this just missing before? (It doesn't seem related to the custom conversion part.)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs [hunk 16]
```diff
@@ -160,7 +161,25 @@ internal object GetValidatedValue()
         /// </summary>
         /// <param name="options">Options to configure the <c>SpannerDbType</c> for this <c>SpannerParameter</c>.</param>
         /// <returns>The configured <see cref="SpannerDbType"/>.</returns>
-        internal SpannerDbType GetConfiguredSpannerDbType(SpannerConversionOptions options) => SpannerDbType;
+        internal SpannerDbType GetConfiguredSpannerDbType(SpannerConversionOptions options)
+        {
+            // Only if SpannerDbType of parameter is not explicitly provided by user.
+            if (_spannerDbType == null)
+            {
+                if (Value is decimal && options != null)
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs:169]: Will options ever be null here? How hard would it be to ensure it's never null?

**Review by jskeet — COMMENTED** [review]
Nearly there, I think. A couple of actual behavioral changes I'd like to see (case handling in SpannerConversionOptions, and the argument exception around PgNumeric) but the rest are potential future changes.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs (hunk 15 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.cs:320]: Yes Jon. It should have gone with SpannerDate changes.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs (hunk 16 above)
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerParameter.cs:169]: I will make the change to ensure options are not null.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs (hunk 8 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:619]: Okay, it sounds like what you *really* want is something saying "is the type code string or float64" - although I may have misunderstood the relationship between the type code and the Value kind here.

Fundamentally this code needs to be much clearer about what's valid to convert into PgNumeric...

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs (hunk 12 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConversionOptions.cs:178]: I would probably do something like having a `Dictionary<string, (SpannerDbType, Type)>` for each option to "source and target types". Then you can find out any options that conflict by whether there multiple options with the same source. (That can also be a case-insensitive dictionary.) I can picture the code, but it's somewhat fiddly to actually describe. But basically we'd end up with the two dictionaries (one in each direction) having almost all the type-specific aspects, so adding a new conversion option should be a small change.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 17]
```diff
@@ -140,6 +145,114 @@ private string ValidatedDataSource(string dataSource)
             return dataSource;
         }
 
+        /// <summary>
+        /// Option to configure the default CLR type to SpannerDbType mapping. This option comes into picture
+        /// only if <see cref="SpannerDbType"/> and <see cref="System.Data.DbType"/> of the <see cref="SpannerParameter"/>
+        /// are not explicitly provided. Currently only <see cref="decimal"/> and <see cref="DateTime"/> CLR types are supported.
+        /// </summary>
+        /// <remarks>
```
[the remark this answers is not in the document]
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:155]: Done.

**Review by Rishabh-V — COMMENTED** [review]
Hi Jon. Have made most of the requested changes. Didn't get the string.Split overload in .netstandard2.0. Have now made case insensitive comparison and PgNumeric and Numeric handling is updated with options. Please take a look.
 
Have added TODO to the following items. 
(Will raise a new PR for them)
1) Validation methods code reuse.
2) Use Dictionary for validating multiple mappings.
3) RunReadStream should use builder correctly in Tests.
4) Docs should be updated in connection string. 

Have squashed the commits to make it easier to review. Once this is good, I can raise two different PRs. One for decimal and Float64 and one for DateTime and Date for merging. Thank you.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs [hunk 18]
```diff
@@ -603,23 +608,36 @@ private object ConvertToClrTypeImpl(Value wireValue, System.Type targetClrType,
...
+                bool isPgNumeric = TypeAnnotationCode == TypeAnnotationCode.PgNumeric;
+                bool isFloat64 = TypeCode == TypeCode.Float64;
+                // Check if Float64 is being read as PgNumeric due to user provided type mapping. 
+                bool configuredFloat64AsPgNumeric = isFloat64 && options.Float64ToConfiguredClrType == typeof(PgNumeric); 
+                // For PgNumeric, the value is contained in Value.StringValue.
+                // If the user configures that Float64 should be mapped to PgNumeric CLR type in connection string,
+                // we will need to read Float64 as PgNumeric. The Float64 value is populated in NumberValue of Value.
+                // If the TypeCode is Float64 due to user configured mapping or TypeAnnotationCode is PgNumeric, it is fine.
+                // In all other cases, throw an exception.
+                if (!isPgNumeric && !configuredFloat64AsPgNumeric)
```
[the remark this answers is not in the document]
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:629]: (I think it's slightly differently wrong now to how it was wrong before... it shouldn't matter whether the way that we've ended up wanting to convert a value to PgNumeric is via connection string options or a direct request.)

**Review by jskeet — APPROVED** [review]
I haven't rechecked the tests, although I've looked through the new SpannerConversionOptions tests. As noted, I'm still definitely uncomfortable about the validation condition - we need to get that sorted before we next release, but I'd like to do it in a standalone fashion.

**Comment by Rishabh-V** [comment]
> I haven't rechecked the tests, although I've looked through the new SpannerConversionOptions tests. As noted, I'm still definitely uncomfortable about the validation condition - we need to get that sorted before we next release, but I'd like to do it in a standalone fashion.

Sure Jon.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs (hunk 18 above)
[the remark this answers is not in the document]
Comment (Rishabh-V) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDbType.ValueConversion.cs:629]: Sure Jon.

**Comment by jskeet** [comment]
Can we now close #5837, #5839, #8463 and #8472? (Possibly wait until we've got docs before closing the issues, actually?)
